File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/spine/ports/Authorization.kt
// ── spine/ports/authorization — the product-owned seam (14.3, 17.1) ────────
// This is where G6's unattended confirmer becomes representable and where its per-tenant budget's
// per-tenant budget belongs. Both are enforced at the BOUNDARY, before the fold,
// and their verdict rides the committed record as an ordered G9 fixture — so a
// re-fold reproduces the decision without calling either of these again.
//
// Note what does NOT happen here: Actor does not grow a variant. The actor
// contract grows only at architecture revision, never per application — and a
// product's tenth kind of confirmer is not a revision. Authority is an opaque id,
// so it adds no type and no case.

package adr.spine.ports

import adr.contract.ToolResult
import adr.spine.pure.Actor
import adr.spine.pure.Authority
import adr.spine.pure.SessionId
import adr.spine.pure.Signature

/**
 * Who is this action being admitted for? Resolved once per step, at the boundary.
 * The agent's own run, a policy tier, a second-agent reviewer, a deferred approval
 * queue and a human host are all just different Authority values.
 */
interface AuthorityResolver {
    fun authorityOf(by: Actor, session: SessionId): Authority
}

/**
 * May THIS authority confirm THIS irreversible action, given that `requestedBy`
 * asked for it? The self-confirm case is already denied structurally by the gate;
 * this seam is where a product adds its own rules on top.
 */
interface ConfirmPolicy {
    fun mayConfirm(sig: Signature, result: ToolResult, requestedBy: Authority): Boolean
}

File: agent-driven-architecture-main/examples/typescript/src/spine/ports/authorization.ts
// ── spine/ports/authorization — the PRODUCT-OWNED seam (14.3, G6) ──────────
// INTERFACES ONLY (C11).
//
// This is where the book already routes actor-keyed checks, and it is where
// G6's unattended confirmer becomes reachable. `authorityOf` answers "under
// whose permission is this stream acting right now" — a policy tier, a
// second-agent reviewer, a deferred approval queue and a human host are all
// just different Authority ids, so no product-owned confirmer ever adds an Actor
// variant — the actor contract grows at architecture revision, never per app.
//
// G6's per-tenant budget has the same home: a folded budget bounds only
// aggregates within the unit of work its stream is scoped to, so anything
// spanning k concurrent sessions is enforced HERE, before the fold, with its
// verdict captured on the committed record like any other decision.

import type { Actor, Authority, Signature } from "../pure/actor";
import type { SessionId } from "../pure/ids";
import type { ToolResultBase } from "../pure/tool-result";

export interface AuthorityResolver {
  authorityOf(by: Actor, session: SessionId): Authority;
}

export interface ConfirmPolicy<S> {
  mayConfirm(sig: Signature, result: ToolResultBase, state: S): boolean;
}

export interface Authorization<S> extends AuthorityResolver, ConfirmPolicy<S> {}
